[WFLY-5564] Unable to set use-journal-store if use-jdbc-store was set…

TomasHofman opened the pull request:
… to true before

https://issues.jboss.org/browse/JBEAP-1354
https://issues.jboss.org/browse/WFLY-5564

Modifications:
- Removed USE_JDBC_STORE.setAlternatives(USE_JOURNAL_STORE) and inverse because it prevented to set the first even if the other was set to false. Implemented special write handler to validate this correctly.
- Removed JDBC_STORE_DATASOURCE.setRequires(USE_JDBC_STORE) and instead implemented special write handler checking that JDBC_STORE_DATASOURCE is set if USE_JDBC_STORE is true.

TomasHofman committed:
[WFLY-5564] Unable to set use-journal-store if use-jdbc-store was set to true before

bstansberry commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/transaction/ObjectStoreTypeTest.java:
```diff
@@ -69,19 +69,19 @@ public void after() throws Exception {
 
     @Test
-    public void testHornetQObjectStore() throws IOException, MgmtOperationException {
+    public void testJournalObjectStore() throws IOException, MgmtOperationException {
```
We need a new test instead of repurposing the existing one. The use-hornetq-store attribute is still part of the management API so we should test what happens when it is used.

bstansberry commented:
TransactionSubsystemAdd should do these validations as well.

TomasHofman commented:
Ah sorry, forgot about that during weekend.

TomasHofman commented:
Thanks for reminding me!

bstansberry commented:
@n1hility You asked about the compatibility impacts of this. I believe the only impacts are:

In WildFly < 10 this would work:

{code}
/subsystem=transactions:write-attribute(name=use-jdbc-store,value=true)
/subsystem=transactions:write-attribute(name=jdbc-store-datasource,value=java:datasources/blah)
{code}

Now it won't. The user would have to encapsulate those two calls in a batch.

The old behavior was incorrect as the first operation by itself puts the config in an invalid state. This PR corrects that by adding appropriate validation.

The business with .setAlternatives() on USE_JOURNAL_STORE and USE_JDBC_STORE won't affect compatibility since prior to WF 10 there was no checking of those constraints. Also, the "alternatives" support in the kernel isn't really appropriate for these attributes, as the constraint is really that both can't be defined with a value of 'true' while the kernel rejects any case where they are both defined, regardless of value. I'd like to enhance the kernel to somehow account for this distinction, but in the meantime it's not correct to say the two attributes are alternatives given the present meaning of the metadata.

TomasHofman commented:
Updated
* test ObjectStoreTypeTest renamed to ObjectStoreTypeTestCase so it's part of normal surefire execution.
* test for hornetq store remains
* added validations to TransactionSubsystemAdd

TomasHofman commented:
Retest this please.

n1hility commented:
retest this please

bstansberry commented:
Looks good!
